<span style="font-family: Arial; font-weight:bold;font-size:1.5em;">Problem Statement

Employee Attrition has always been a crucial concern for a company. Attrition means a loss to the company in terms of cost invested in training an employee. There are a few companies which were forced to close due to high level of attrition. To minimize attrition several companies tries to analyze and understand several factors which migh lead to an employee to leave the company. 
The IBM HR Attrition Case Study is a dataset which aims to identify important factors that might be influential in determining an employee who might leave the company. You are hired by the company and your objective is to undestand the data and to identify the patterns in the data which will help HR team to take decisive actions to reduce the attrition rate.

There are two given dataset, Company_Data.csv contains information of an employee related to the company and Personal_Data cotains personal information of an employee.

## Data Dictionary

1. EmpId-              Employee ID
2. AGE-                Employee Age
3. EducationField-     Field of Education
4. Gender        -     Employee Gender
5. MaritalStatus -     Employee Marital Status
6. DistanceFromHome-   Distance from home to work (in miles)
7. BusinessTravel-     Frequency of business travel
8. Department-         Department of the employee 
9. HourlyRate-         Hourly rate of an employee
10. JobLevel-          Job Level of the employee. Higher job level means senior post
11. JobRole-           Job Role of an employee
12. JobSatisfaction-   Job Satisfaction level of  an employee
13. MonthlyRate-       Montly rate of an employee
14. NumCompaniesWorked-How many companies has the employee worked before
15. OverTime-          Does the employee gets overtime
16. PercentSalaryHike- Average percentage take for the employee
17. TotalWorkingYears- Total Years of Experience of an employee
18. WorkLifeBalance-   Work life balance of an employee (4 means highly satisfied)
19. YearsAtCompany-    Years employee have worked in current company
20. YearsInCurrentRole-Years employee have worked in current role
21. TotalWorkingYears- Total Years of Experience of an employee
22. YearsSinceLastPromotion-   Number of years since last promotion
23. YearsWithCurrManager- Number of years with current manager
24. Attrition-         Employee has left the company or not


### Change your Working Directory

In [1]:
import os

In [3]:
print(os.name)

nt


In [4]:
print(os.getcwd()) # Got current working directory

C:\Users\anime\Downloads\Great Learning\Introduction to Python\Week 1 and 2 Practice Assessment


In [5]:
os.listdir('.') # gives the files in the current working directory

['.ipynb_checkpoints',
 'Company_Data.csv',
 'Personal_Data.csv',
 'Questions_HR_Attrition.ipynb',
 'Questions_Store_Sale.ipynb',
 'sales.csv',
 'Solution_Forbes_Billionaires.html',
 'Solution_Forbes_Billionaires.pdf',
 'Solution_HR Attrition.pdf',
 'Solution_HR+Attrition.html',
 'Solution_Store Sales.pdf',
 'Solution_Store+Sales.html']

In [6]:
os.mkdir('C:\\Users\\anime\\Downloads\\Great Learning\\\Introduction to Python\\Week 1 and 2 Practice Assessment\\new_dir')

In [8]:
# Answer
os.chdir('C:\\Users\\anime\\Downloads\\Great Learning\\\Introduction to Python\\Week 1 and 2 Practice Assessment\\new_dir')

In [9]:
os.getcwd()

'C:\\Users\\anime\\Downloads\\Great Learning\\Introduction to Python\\Week 1 and 2 Practice Assessment\\new_dir'

### Q1. Import all the necessary libraries (Numpy and Pandas)

In [11]:
import numpy as np
import pandas as pd

### Q2. Import both csv files (and name Company_Data as "company" and Personal_Data as "personal" )

Load the Data file into Python DataFrame using pandas read_csv method

In [12]:
company = pd.read_csv('Company_Data.csv')
personal = pd.read_csv('Personal_Data.csv')

### Q3.  View Top 10 rows of both the datasets

Hint:Head function is used to view the top records. The number of records to be viewed, needs to be  given in the parenthesis. Default value is 5.

In [17]:
company.head(10)

EmpID     BusinessTravel              Department  HourlyRate  JobLevel  \
0  2102210      Travel_Rarely                   Sales          94       2.0   
1  2102211  Travel_Frequently  Research & Development          61       2.0   
2  2102212      Travel_Rarely  Research & Development          92       1.0   
3  2102213  Travel_Frequently  Research & Development          56       1.0   
4  2102214      Travel_Rarely  Research & Development          40       1.0   
5  2102215  Travel_Frequently  Research & Development          79       1.0   
6  2102216      Travel_Rarely  Research & Development          81       1.0   
7  2102217      Travel_Rarely  Research & Development          67       1.0   
8  2102218  Travel_Frequently  Research & Development          44       3.0   
9  2102219      Travel_Rarely  Research & Development          94       2.0   

                     JobRole  JobSatisfaction  MonthlyRate  \
0            Sales Executive                4        19479   
1         Research Scientist                2        24907   
2      Laboratory Technician                3         2396   
3         Research Scientist                3        23159   
4      Laboratory Technician                2        16632   
5      Laboratory Technician                4        11864   
6      Laboratory Technician                1         9964   
7      Laboratory Technician                3        13335   
8     Manufacturing Director                3         8787   
9  Healthcare Representative                3        16577   

   NumCompaniesWorked OverTime  PercentSalaryHike  TotalWorkingYears  \
0                   8      Yes                 11                  8   
1                   1       No                 23                 10   
2                   6      Yes                 15                  7   
3                   1      Yes                 11                  8   
4                   9       No                 12                  6   
5                   0       No                 13                  8   
6                   4      Yes                 20                 12   
7                   1       No                 22                  1   
8                   0       No                 21                 10   
9                   6       No                 13                 17   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0              1.0               6                   4   
1              3.0              10                   7   
2              3.0               0                   0   
3              3.0               8                   7   
4              3.0               2                   2   
5              2.0               7                   7   
6              2.0               1                   0   
7              3.0               1                   0   
8              3.0               9                   7   
9              2.0               7                   7   

   YearsSinceLastPromotion  YearsWithCurrManager Attrition  
0                        0                     5       Yes  
1                        1                     7        No  
2                        0                     0       Yes  
3                        3                     0        No  
4                        2                     2        No  
5                        3                     6        No  
6                        0                     0        No  
7                        0                     0        No  
8                        1                     8        No  
9                        7                     7        No

In [16]:
personal.head(10)

EmpID  Age EducationField  Gender MaritalStatus  DistanceFromHome
0  2102210   41  Life Sciences  Female        Single                 1
1  2102211   49  Life Sciences    Male       Married                 8
2  2102212   37          Other    Male        Single                 2
3  2102213   33  Life Sciences  Female       Married                 3
4  2102214   27        Medical    Male       Married                 2
5  2102215   32  Life Sciences    Male        Single                 2
6  2102216   59        Medical  Female       Married                 3
7  2102217   30  Life Sciences    Male      Divorced                24
8  2102218   38  Life Sciences    Male        Single                23
9  2102219   36        Medical    Male       Married                27

### Q4. Merge both datasets and name the merged dataset as df

Hint:Merge the datasets using pd.merge() command- Use how="outer" and on="EmpID"

In [19]:
df = pd.merge(company,personal, how = "outer", on = "EmpID")
df

EmpID     BusinessTravel              Department  HourlyRate  \
0     2102210      Travel_Rarely                   Sales          94   
1     2102211  Travel_Frequently  Research & Development          61   
2     2102212      Travel_Rarely  Research & Development          92   
3     2102213  Travel_Frequently  Research & Development          56   
4     2102214      Travel_Rarely  Research & Development          40   
5     2102215  Travel_Frequently  Research & Development          79   
6     2102216      Travel_Rarely  Research & Development          81   
7     2102217      Travel_Rarely  Research & Development          67   
8     2102218  Travel_Frequently  Research & Development          44   
9     2102219      Travel_Rarely  Research & Development          94   
10    2102220      Travel_Rarely  Research & Development          84   
11    2102221      Travel_Rarely  Research & Development          49   
12    2102222      Travel_Rarely  Research & Development          31   
13    2102223      Travel_Rarely  Research & Development          93   
14    2102224      Travel_Rarely  Research & Development          50   
15    2102225      Travel_Rarely  Research & Development          51   
16    2102226      Travel_Rarely  Research & Development          80   
17    2102227         Non-Travel  Research & Development          96   
18    2102228      Travel_Rarely                   Sales          78   
19    2102229      Travel_Rarely  Research & Development          45   
20    2102230         Non-Travel  Research & Development          96   
21    2102231      Travel_Rarely                   Sales          82   
22    2102232      Travel_Rarely  Research & Development          53   
23    2102233      Travel_Rarely  Research & Development          96   
24    2102234      Travel_Rarely  Research & Development          83   
25    2102235      Travel_Rarely  Research & Development          58   
26    2102236  Travel_Frequently  Research & Development          72   
27    2102237      Travel_Rarely                   Sales          48   
28    2102238      Travel_Rarely  Research & Development          42   
29    2102239      Travel_Rarely                   Sales          83   
...       ...                ...                     ...         ...   
1440  2103650  Travel_Frequently  Research & Development          97   
1441  2103651         Non-Travel  Research & Development          57   
1442  2103652      Travel_Rarely  Research & Development          36   
1443  2103653      Travel_Rarely  Research & Development          56   
1444  2103654      Travel_Rarely  Research & Development          72   
1445  2103655      Travel_Rarely  Research & Development          60   
1446  2103656      Travel_Rarely                   Sales          95   
1447  2103657         Non-Travel                   Sales          88   
1448  2103658      Travel_Rarely                   Sales          57   
1449  2103659      Travel_Rarely  Research & Development          78   
1450  2103660      Travel_Rarely         Human Resources          31   
1451  2103661      Travel_Rarely                   Sales         100   
1452  2103662  Travel_Frequently                   Sales          94   
1453  2103663      Travel_Rarely                   Sales         100   
1454  2103664      Travel_Rarely                   Sales          50   
1455  2103665      Travel_Rarely  Research & Development          52   
1456  2103666  Travel_Frequently  Research & Development          80   
1457  2103667      Travel_Rarely  Research & Development          98   
1458  2103668      Travel_Rarely  Research & Development          62   
1459  2103669      Travel_Rarely  Research & Development          46   
1460  2103670      Travel_Rarely  Research & Development          73   
1461  2103671      Travel_Rarely                   Sales          39   
1462  2103672      Travel_Rarely                   Sales          60   
1463  2103673         Non-Travel  Research & Dev

 ### Q5. View Last 5 records of combined dataframe "df".

Hint:Tail function is used to view the last records. The number of records to be viewed, needs to be given in the parenthesis.Default value is 5.

In [20]:
df.tail()

EmpID     BusinessTravel              Department  HourlyRate  \
1465  2103675  Travel_Frequently  Research & Development          41   
1466  2103676      Travel_Rarely  Research & Development          42   
1467  2103677      Travel_Rarely  Research & Development          87   
1468  2103678  Travel_Frequently                   Sales          63   
1469  2103679      Travel_Rarely  Research & Development          82   

      JobLevel                    JobRole  JobSatisfaction  MonthlyRate  \
1465       2.0      Laboratory Technician                4        12290   
1466       3.0  Healthcare Representative                1        21457   
1467       2.0     Manufacturing Director                2         5174   
1468       2.0            Sales Executive                2        13243   
1469       2.0      Laboratory Technician                3        10228   

      NumCompaniesWorked OverTime       ...         YearsAtCompany  \
1465                   4       No       ...                      5   
1466                   4       No       ...                      7   
1467                   1      Yes       ...                      6   
1468                   2       No       ...                      9   
1469                   2       No       ...                      4   

      YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
1465                   2                        0                     3   
1466                   7                        1                     7   
1467                   2                        0                     3   
1468                   6                        0                     8   
1469                   3                        1                     2   

      Attrition  Age  EducationField Gender  MaritalStatus DistanceFromHome  
1465         No   36         Medical   Male        Married               23  
1466         No   39         Medical   Male        Married                6  
1467         No   27   Life Sciences   Male        Married                4  
1468         No   49         Medical   Male        Married                2  
1469         No   34         Medical   Male        Married                8  

[5 rows x 23 columns]

### Q6. Display the shape/dimensions of dataset.

Displaying the shape of the Data Frame in which first value gives number of Rows and second value gives number of columns.

Hint: Use "shape" function 

In [21]:
# Answer
df.shape

(1470, 23)

In [27]:
print("The number of rows are: ", df.shape[0], "\nThe number of columns are: ", df.shape[1])

The number of rows are:  1470 
The number of columns are:  23


### Q7. Check for missing values. Display number of null values per column.

Hint: use isnull() function

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1470 entries, 0 to 1469
Data columns (total 23 columns):
EmpID                      1470 non-null int64
BusinessTravel             1470 non-null object
Department                 1470 non-null object
HourlyRate                 1470 non-null int64
JobLevel                   1464 non-null float64
JobRole                    1470 non-null object
JobSatisfaction            1470 non-null int64
MonthlyRate                1470 non-null int64
NumCompaniesWorked         1470 non-null int64
OverTime                   1470 non-null object
PercentSalaryHike          1470 non-null int64
TotalWorkingYears          1470 non-null int64
WorkLifeBalance            1466 non-null float64
YearsAtCompany             1470 non-null int64
YearsInCurrentRole         1470 non-null int64
YearsSinceLastPromotion    1470 non-null int64
YearsWithCurrManager       1470 non-null int64
Attrition                  1470 non-null object
Age                        1470 non-nu

In [31]:
# Answer
df.isnull().sum()

EmpID                      0
BusinessTravel             0
Department                 0
HourlyRate                 0
JobLevel                   6
JobRole                    0
JobSatisfaction            0
MonthlyRate                0
NumCompaniesWorked         0
OverTime                   0
PercentSalaryHike          0
TotalWorkingYears          0
WorkLifeBalance            4
YearsAtCompany             0
YearsInCurrentRole         0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
Attrition                  0
Age                        0
EducationField             0
Gender                     0
MaritalStatus              0
DistanceFromHome           0
dtype: int64

In [32]:
type(df.isnull().sum())

pandas.core.series.Series

There are missing values present in the "Joblevel" and "WorklifeBalance" variables of the dataset.

### Q8.  Drop null values from "df" and check if there are any missing value left after dropping null values

In [37]:
df.dropna(inplace = True)
df

EmpID     BusinessTravel              Department  HourlyRate  \
0     2102210      Travel_Rarely                   Sales          94   
1     2102211  Travel_Frequently  Research & Development          61   
2     2102212      Travel_Rarely  Research & Development          92   
3     2102213  Travel_Frequently  Research & Development          56   
4     2102214      Travel_Rarely  Research & Development          40   
5     2102215  Travel_Frequently  Research & Development          79   
6     2102216      Travel_Rarely  Research & Development          81   
7     2102217      Travel_Rarely  Research & Development          67   
8     2102218  Travel_Frequently  Research & Development          44   
9     2102219      Travel_Rarely  Research & Development          94   
10    2102220      Travel_Rarely  Research & Development          84   
15    2102225      Travel_Rarely  Research & Development          51   
16    2102226      Travel_Rarely  Research & Development          80   
17    2102227         Non-Travel  Research & Development          96   
18    2102228      Travel_Rarely                   Sales          78   
19    2102229      Travel_Rarely  Research & Development          45   
20    2102230         Non-Travel  Research & Development          96   
21    2102231      Travel_Rarely                   Sales          82   
22    2102232      Travel_Rarely  Research & Development          53   
23    2102233      Travel_Rarely  Research & Development          96   
24    2102234      Travel_Rarely  Research & Development          83   
25    2102235      Travel_Rarely  Research & Development          58   
26    2102236  Travel_Frequently  Research & Development          72   
27    2102237      Travel_Rarely                   Sales          48   
28    2102238      Travel_Rarely  Research & Development          42   
29    2102239      Travel_Rarely                   Sales          83   
30    2102240      Travel_Rarely  Research & Development          78   
31    2102241      Travel_Rarely  Research & Development          41   
32    2102242      Travel_Rarely  Research & Development          83   
33    2102243      Travel_Rarely                   Sales          56   
...       ...                ...                     ...         ...   
1440  2103650  Travel_Frequently  Research & Development          97   
1441  2103651         Non-Travel  Research & Development          57   
1442  2103652      Travel_Rarely  Research & Development          36   
1443  2103653      Travel_Rarely  Research & Development          56   
1444  2103654      Travel_Rarely  Research & Development          72   
1445  2103655      Travel_Rarely  Research & Development          60   
1446  2103656      Travel_Rarely                   Sales          95   
1447  2103657         Non-Travel                   Sales          88   
1448  2103658      Travel_Rarely                   Sales          57   
1449  2103659      Travel_Rarely  Research & Development          78   
1450  2103660      Travel_Rarely         Human Resources          31   
1451  2103661      Travel_Rarely                   Sales         100   
1452  2103662  Travel_Frequently                   Sales          94   
1453  2103663      Travel_Rarely                   Sales         100   
1454  2103664      Travel_Rarely                   Sales          50   
1455  2103665      Travel_Rarely  Research & Development          52   
1456  2103666  Travel_Frequently  Research & Development          80   
1457  2103667      Travel_Rarely  Research & Development          98   
1458  2103668      Travel_Rarely  Research & Development          62   
1459  2103669      Travel_Rarely  Research & Development          46   
1460  2103670      Travel_Rarely  Research & Development          73   
1461  2103671      Travel_Rarely                   Sales          39   
1462  2103672      Travel_Rarely                   Sales          60   
1463  2103673         Non-Travel  Research & Dev

In [39]:
df.isnull().sum()

EmpID                      0
BusinessTravel             0
Department                 0
HourlyRate                 0
JobLevel                   0
JobRole                    0
JobSatisfaction            0
MonthlyRate                0
NumCompaniesWorked         0
OverTime                   0
PercentSalaryHike          0
TotalWorkingYears          0
WorkLifeBalance            0
YearsAtCompany             0
YearsInCurrentRole         0
YearsSinceLastPromotion    0
YearsWithCurrManager       0
Attrition                  0
Age                        0
EducationField             0
Gender                     0
MaritalStatus              0
DistanceFromHome           0
dtype: int64

### Q9. Check the datatype, number of non negative values and name of each variable in the dataset.

Hint : Use info() function

In [40]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1460 entries, 0 to 1469
Data columns (total 23 columns):
EmpID                      1460 non-null int64
BusinessTravel             1460 non-null object
Department                 1460 non-null object
HourlyRate                 1460 non-null int64
JobLevel                   1460 non-null float64
JobRole                    1460 non-null object
JobSatisfaction            1460 non-null int64
MonthlyRate                1460 non-null int64
NumCompaniesWorked         1460 non-null int64
OverTime                   1460 non-null object
PercentSalaryHike          1460 non-null int64
TotalWorkingYears          1460 non-null int64
WorkLifeBalance            1460 non-null float64
YearsAtCompany             1460 non-null int64
YearsInCurrentRole         1460 non-null int64
YearsSinceLastPromotion    1460 non-null int64
YearsWithCurrManager       1460 non-null int64
Attrition                  1460 non-null object
Age                        1460 non-nu

### Q10. Check the summary of the dataset
Hint: To see the summary of a dataframe we will use describe function.

In [42]:
df.describe(include = "all") # include="all" will give summary of all the categorical and numerical variables

EmpID BusinessTravel              Department   HourlyRate  \
count   1.460000e+03           1460                    1460  1460.000000   
unique           NaN              3                       3          NaN   
top              NaN  Travel_Rarely  Research & Development          NaN   
freq             NaN           1036                     952          NaN   
mean    2.102949e+06            NaN                     NaN    65.914384   
std     4.220800e+02            NaN                     NaN    20.316655   
min     2.102210e+06            NaN                     NaN    30.000000   
25%     2.102585e+06            NaN                     NaN    48.000000   
50%     2.102950e+06            NaN                     NaN    66.000000   
75%     2.103314e+06            NaN                     NaN    84.000000   
max     2.103679e+06            NaN                     NaN   100.000000   

           JobLevel          JobRole  JobSatisfaction   MonthlyRate  \
count   1460.000000             1460      1460.000000   1460.000000   
unique          NaN                9              NaN           NaN   
top             NaN  Sales Executive              NaN           NaN   
freq            NaN              325              NaN           NaN   
mean       2.065753              NaN         2.729452  14299.465068   
std        1.107544              NaN         1.103963   7123.418042   
min        1.000000              NaN         1.000000   2094.000000   
25%        1.000000              NaN         2.000000   8039.750000   
50%        2.000000              NaN         3.000000  14225.500000   
75%        3.000000              NaN         4.000000  20460.500000   
max        5.000000              NaN         4.000000  26999.000000   

        NumCompaniesWorked OverTime       ...         YearsAtCompany  \
count          1460.000000     1460       ...            1460.000000   
unique                 NaN        2       ...                    NaN   
top                    NaN       No       ...                    NaN   
freq                   NaN     1048       ...                    NaN   
mean              2.696575      NaN       ...               7.028767   
std               2.499154      NaN       ...               6.139390   
min               0.000000      NaN       ...               0.000000   
25%               1.000000      NaN       ...               3.000000   
50%               2.000000      NaN       ...               5.000000   
75%               4.000000      NaN       ...              10.000000   
max               9.000000      NaN       ...              40.000000   

        YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
count          1460.000000              1460.000000           1460.000000   
unique                 NaN                      NaN                   NaN   
top                    NaN                      NaN                   NaN   
freq                   NaN                      NaN                   NaN   
mean              4.243836                 2.197945              4.133562   
std               3.628309                 3.229596              3.573852   
min               0.000000                 0.000000              0.000000   
25%               2.000000                 0.000000              2.000000   
50%               3.000000                 1.000000              3.000000   
75%               7.000000                 3.000000              7.000000   
max              18.000000                15.000000             17.000000   

        Attrition          Age  EducationField Gender  MaritalStatus  \
count        1460  1460.000000            1460   1460           1460   
unique          2          NaN               6      2              3   
top            No          NaN   Life Sciences   Male        Married   
freq         1225          NaN             600    875            671   
mean          NaN    36.910959             NaN    NaN            NaN   
std           NaN     9.126

### Q11. Drop  the variable EmpID from the dataframe df.
To drop any variable, use drop() function.

In [44]:
df.drop("EmpID", axis =1, inplace = True)

In [45]:
df.head()

BusinessTravel              Department  HourlyRate  JobLevel  \
0      Travel_Rarely                   Sales          94       2.0   
1  Travel_Frequently  Research & Development          61       2.0   
2      Travel_Rarely  Research & Development          92       1.0   
3  Travel_Frequently  Research & Development          56       1.0   
4      Travel_Rarely  Research & Development          40       1.0   

                 JobRole  JobSatisfaction  MonthlyRate  NumCompaniesWorked  \
0        Sales Executive                4        19479                   8   
1     Research Scientist                2        24907                   1   
2  Laboratory Technician                3         2396                   6   
3     Research Scientist                3        23159                   1   
4  Laboratory Technician                2        16632                   9   

  OverTime  PercentSalaryHike       ...         YearsAtCompany  \
0      Yes                 11       ...                      6   
1       No                 23       ...                     10   
2      Yes                 15       ...                      0   
3      Yes                 11       ...                      8   
4       No                 12       ...                      2   

   YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
0                   4                        0                     5   
1                   7                        1                     7   
2                   0                        0                     0   
3                   7                        3                     0   
4                   2                        2                     2   

   Attrition  Age EducationField  Gender MaritalStatus DistanceFromHome  
0        Yes   41  Life Sciences  Female        Single                1  
1         No   49  Life Sciences    Male       Married                8  
2        Yes   37          Other    Male        Single                2  
3         No   33  Life Sciences  Female       Married                3  
4         No   27        Medical    Male       Married                2  

[5 rows x 22 columns]

### Q12. Check how many employees left the company from the given dataset
Hint :Use value_counts() function

In [48]:
df["Attrition"].value_counts()

No     1225
Yes     235
Name: Attrition, dtype: int64

In [51]:
# or
df.Attrition.value_counts()

No     1225
Yes     235
Name: Attrition, dtype: int64

In the given dataset, 235 employees left the company.

In [52]:
df.Attrition.value_counts(normalize=True)
# normalize=True gives the output in the ratio terms

No     0.839041
Yes    0.160959
Name: Attrition, dtype: float64

In the given dataset, 16.09% of the employees left the company.

### Q13. Create a subset "df_new" from the dataset "df" which contain data of all the employees who have attrited. Also check the shape of the new dataset

In [56]:
df_new = df[df["Attrition"] == "Yes"]
df_new.head()

BusinessTravel              Department  HourlyRate  JobLevel  \
0       Travel_Rarely                   Sales          94       2.0   
2       Travel_Rarely  Research & Development          92       1.0   
21      Travel_Rarely                   Sales          82       1.0   
24      Travel_Rarely  Research & Development          83       1.0   
26  Travel_Frequently  Research & Development          72       1.0   

                  JobRole  JobSatisfaction  MonthlyRate  NumCompaniesWorked  \
0         Sales Executive                4        19479                   8   
2   Laboratory Technician                3         2396                   6   
21   Sales Representative                1         6986                   7   
24     Research Scientist                1        17102                   2   
26     Research Scientist                1         4681                   1   

   OverTime  PercentSalaryHike       ...         YearsAtCompany  \
0       Yes                 11       ...                      6   
2       Yes                 15       ...                      0   
21       No                 23       ...                      5   
24       No                 11       ...                      4   
26      Yes                 22       ...                     10   

    YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  \
0                    4                        0                     5   
2                    0                        0                     0   
21                   3                        0                     3   
24                   2                        1                     3   
26                   2                        6                     7   

    Attrition  Age EducationField  Gender MaritalStatus DistanceFromHome  
0         Yes   41  Life Sciences  Female        Single                1  
2         Yes   37          Other    Male        Single                2  
21        Yes   36  Life Sciences    Male        Single                9  
24        Yes   34        Medical    Male        Single                6  
26        Yes   32  Life Sciences  Female        Single               16  

[5 rows x 22 columns]

In [57]:
df_new.shape

(235, 22)

### Q14. Create the top 5 rows and first 5 columns in a matrix format. Use 

---

dataframe "df"

Hint- use iloc function

In [58]:
df.iloc[:5,:5]

BusinessTravel              Department  HourlyRate  JobLevel  \
0      Travel_Rarely                   Sales          94       2.0   
1  Travel_Frequently  Research & Development          61       2.0   
2      Travel_Rarely  Research & Development          92       1.0   
3  Travel_Frequently  Research & Development          56       1.0   
4      Travel_Rarely  Research & Development          40       1.0   

                 JobRole  
0        Sales Executive  
1     Research Scientist  
2  Laboratory Technician  
3     Research Scientist  
4  Laboratory Technician

### Q15.  Analyze the mean for all the variables for employees who have attrited and whose who have not.

Hint: Use groupby() function

In [61]:
df.groupby(["Attrition"]).mean()

HourlyRate  JobLevel  JobSatisfaction   MonthlyRate  \
Attrition                                                        
No          65.977959  2.146939         2.780408  14254.193469   
Yes         65.582979  1.642553         2.463830  14535.455319   

           NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  \
Attrition                                                             
No                   2.648980          15.242449          11.862857   
Yes                  2.944681          15.093617           8.280851   

           WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
Attrition                                                        
No                2.782041        7.388571            4.497959   
Yes               2.659574        5.153191            2.919149   

           YearsSinceLastPromotion  YearsWithCurrManager        Age  \
Attrition                                                             
No                        2.243265              4.377143  37.542041   
Yes                       1.961702              2.863830  33.621277   

           DistanceFromHome  
Attrition                    
No                 8.902041  
Yes               10.582979

In [64]:
# or
df.groupby("Attrition").mean()

HourlyRate  JobLevel  JobSatisfaction   MonthlyRate  \
Attrition                                                        
No          65.977959  2.146939         2.780408  14254.193469   
Yes         65.582979  1.642553         2.463830  14535.455319   

           NumCompaniesWorked  PercentSalaryHike  TotalWorkingYears  \
Attrition                                                             
No                   2.648980          15.242449          11.862857   
Yes                  2.944681          15.093617           8.280851   

           WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
Attrition                                                        
No                2.782041        7.388571            4.497959   
Yes               2.659574        5.153191            2.919149   

           YearsSinceLastPromotion  YearsWithCurrManager        Age  \
Attrition                                                             
No                        2.243265              4.377143  37.542041   
Yes                       1.961702              2.863830  33.621277   

           DistanceFromHome  
Attrition                    
No                 8.902041  
Yes               10.582979

### Q16. Check the details of 5 employees who have spent maximum years in the company 

In [65]:
df.sort_values(by = 'YearsAtCompany', ascending = 'False').head(5)

BusinessTravel              Department  HourlyRate  JobLevel  \
727          Non-Travel  Research & Development          73       1.0   
936   Travel_Frequently  Research & Development          83       5.0   
1237      Travel_Rarely                   Sales          34       2.0   
1074      Travel_Rarely  Research & Development          69       2.0   
1287      Travel_Rarely  Research & Development          95       2.0   

                        JobRole  JobSatisfaction  MonthlyRate  \
727          Research Scientist                4        13493   
936                     Manager                2        13035   
1237            Sales Executive                2        12147   
1074  Healthcare Representative                3        22049   
1287  Healthcare Representative                1         5696   

      NumCompaniesWorked OverTime  PercentSalaryHike       ...         \
727                    1       No                 15       ...          
936                    3       No                 22       ...          
1237                   6       No                 15       ...          
1074                   2      Yes                 17       ...          
1287                   5       No                 18       ...          

      YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
727                0                   0                        0   
936                0                   0                        0   
1237               0                   0                        0   
1074               0                   0                        0   
1287               0                   0                        0   

      YearsWithCurrManager  Attrition  Age EducationField  Gender  \
727                      0         No   18  Life Sciences    Male   
936                      0         No   45        Medical  Female   
1237                     0        Yes   32  Life Sciences    Male   
1074                     0         No   33  Life Sciences    Male   
1287                     0         No   42        Medical    Male   

     MaritalStatus DistanceFromHome  
727         Single                5  
936         Single               25  
1237        Single                1  
1074        Single                8  
1287       Married               13  

[5 rows x 22 columns]

### Q17.  Find out the minimum "Percentage Salary Hike" the company has offered?


In [66]:
df['PercentSalaryHike'].min()

11

In [67]:
# or
df.PercentSalaryHike.min()

11

### Q18. What is the maximum Total Working Years of an employee?

In [68]:
df['YearsAtCompany'].max()

40

### Q19. What percentage of employees stay in the company given they have spend more than 7 years with the current manager?

In [72]:
df[df['YearsWithCurrManager'] > 7]['Attrition'].value_counts(normalize = 'True')

No     0.918519
Yes    0.081481
Name: Attrition, dtype: float64

### Q20.  What percentage of employees lives >10 miles from the company?

In [73]:
df[df['DistanceFromHome'] > 10]['DistanceFromHome'].count()

439

In [75]:
df.shape[0]

1460

In [77]:
(df[df['DistanceFromHome'] > 10]['DistanceFromHome'].count()/df.shape[0])*100

30.068493150684933

### Q21. Create a copy of the dataset , remove the "Attrition" variable from the new dataset and check the shape of the dataset

In [78]:
df_copy = df.copy()

In [79]:
df_copy.drop("Attrition", axis = 1, inplace = True)
df_copy.shape

(1460, 21)

# END